### PR TITLE
`export` extensions.

### PR DESCRIPTION
**Pull request by rsmmr**
Addresses #3218 and #3219 .

Documentation in https://github.com/zeek/spicy/pull/1497.

### CODE REVIEW

**Review by bbannier — CHANGES_REQUESTED** [review]

File: testing/btest/spicy/export-type-with-fields.zeek [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+protocol analyzer FOO over TCP:
+    parse with foo::X,
+    port 80/tcp;
+
+export foo::X with { x };
+export foo::X as foo::X1;
+export foo::X as foo::X2 &log;
+export foo::X as foo::X3 with { x, z &log };
+export foo::X as foo::X4 without { x, y };
```
[the remark this answers is not in the document]
Comment (bbannier) [reply testing/btest/spicy/export-type-with-fields.zeek:30]: Interesting idea on the rename syntax, if we swapped the order this should even be possible as an extension of the original syntax in this PR, e.g.,

```zeek
export foo::X as foo::X3 with {
  a from x,
  b &log from z
};
```

This syntax doesn't require dropping `without`. I am still wondering whether the asymmetric attribute support will feel in practice, but I see your point about huge units (though there is also benefit in low-boilerplate explicitness).

[the remark this answers is not in the document]
Comment (rsmmr) [reply testing/btest/spicy/export-type-with-fields.zeek:30]: I'm with Jan here, `without` seems quite useful to have. It doesn't need to be fully symmetric IMO, I see it as a convenience shortcut. I also like the field renaming syntax `x as a`, would be easy to add (but not doing this now).

File: src/spicy/spicyz/glue-compiler.cc [hunk 2]
```diff
@@ -528,6 +517,70 @@ bool GlueCompiler::loadEvtFile(hilti::rt::filesystem::path& path) {
...
+    if ( export_->includes.empty() ) {
+        // Include unless explicitly excluded.
+        if ( export_->excludes.find(field_id) != export_->excludes.end() )
+            field.skip = true;
+    }
+    else {
+        // Exclude unless explicitly included.
+        if ( export_->includes.find(field_id) == export_->includes.end() )
+            field.skip = true;
+    }
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply src/spicy/spicyz/glue-compiler.cc:544]: Not sure about this. The code follows the other cases where the structs capture what the syntax provides, but don't interpret it further upfront. I'll rename `includes` to `with` and `excludes` to `without`, that should make it clearer.

File: src/spicy/runtime-support.h [hunk 3]
```diff
@@ -151,7 +151,7 @@ extern TypePtr create_enum_type(
     const std::string& ns, const std::string& id,
     const hilti::rt::Vector<std::tuple<std::string, hilti::rt::integer::safe<int64_t>>>& labels);
 
-using RecordField = std::tuple<std::string, TypePtr, hilti::rt::Bool>; // (ID, type, optional)
+using RecordField = std::tuple<std::string, TypePtr, hilti::rt::Bool, hilti::rt::Bool>; // (ID, type, optional, &log)
```
[the remark this answers is not in the document]
Comment (rsmmr) [reply src/spicy/runtime-support.h:154]: Ack, I changing that in a new commit (but it doesn't work to use the same struct everywhere).

**Review by bbannier — APPROVED** [review]
🚀

**Comment by awelzel** [comment]
Should this have a NEWS entry? It's a pretty nice enhancement.

**Pull request merged**